cdap-ui/app/cdap/components/DataPrepConnections/SpannerConnection/index.tsx

```diff
@@ -0,0 +1,390 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import * as React from 'react';
+import { Modal, ModalHeader, ModalBody, ModalFooter } from 'reactstrap';
+import {getCurrentNamespace} from 'services/NamespaceStore';
+import T from 'i18n-react';
+import LoadingSVG from 'components/LoadingSVG';
+import MyDataPrepApi from 'api/dataprep';
+import CardActionFeedback, {CARD_ACTION_TYPES} from 'components/CardActionFeedback';
+import {objectQuery} from 'services/helpers';
+import BtnWithLoading from 'components/BtnWithLoading';
+import {ConnectionType} from 'components/DataPrepConnections/ConnectionType';
+
+const PREFIX = 'features.DataPrepConnections.AddConnections.Spanner';
+const ADDCONN_PREFIX = 'features.DataPrepConnections.AddConnections';
+
+const LABEL_COL_CLASS = 'col-xs-3 col-form-label text-xs-right';
+const INPUT_COL_CLASS = 'col-xs-8';
+
+require('./SpannerConnection.scss');
+
+enum ConnectionMode {
+  Add = 'ADD',
+  Edit = 'EDIT',
+  Duplicate = 'DUPLICATE',
+}
+
+interface ISpannerConnectionProps {
+  close: () => void;
+  onAdd: () => void;
+  mode: ConnectionMode;
+  connectionId: string;
+}
+
+interface ISpannerConnectionState {
+  error?: string | object | null;
+  name?: string;
```

Comment:
Do we need to make all the properties optional?

Reply:
I had to do this, otherwise the `setState({[key]: ...})` in the `handleChange` function would complain. Do you know another way around this?